{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "private_outputs": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "#@title Usage instructions\n",
        "\n",
        "#@markdown Enter your gene IDs and choose the tagging terminus, then select `Runtime > Run all` or press `Ctrl+F9` to run all of the code.\n",
        "#@markdown To do further primer design, you only need to re-run the \"User settings\" and \"Find tagging primers\" sections.\n",
        "\n",
        "#@ Sequence retrieval from TriTrypDB, using URL API for sequence retrieval tool.\n",
        "\n",
        "from urllib.request import urlopen\n",
        "import math\n",
        "\n",
        "get_sequence_cache = {}\n",
        "\n",
        "def sequence_request(id: str, upAnchor: str, upSign: str, upOffs: int, downAnchor: str, downSign: str, downOffs: int) -> str:\n",
        "  \"\"\"\n",
        "  Fetch a sequence from TriTrypDB, behaviour as for sequence retrieval tool at https://tritrypdb.org/tritrypdb/app/fasta-tool/gene\n",
        "\n",
        "  :param id: Gene id\n",
        "  :param upAnchor: Upstream anchor: `\"Start\"` transcript start, `\"CodeStart\"` CDS start, `\"CodeEnd\"` CDS end, or `\"End\"` transcript end\n",
        "  :param upSign: Upstream offset direction: `\"plus\"` (downstream) or `\"minus\"` (upstream)\n",
        "  :param upOffs: Upstream offset distance in bases.\n",
        "  :param downAnchor: Downstream anchor: options as for `upAnchor`\n",
        "  :param downSign: Downstream offset direction: options as for `upSign`\n",
        "  :param downOffs: Downstream offest distances in bases.\n",
        "  :return: Sequence at the requested locus\n",
        "  \"\"\"\n",
        "  # construct url\n",
        "  url = \"https://veupathdb.org/cgi-bin/geneSrt?project_id=EuPathDB\"\n",
        "  url += \"&type=genomic\" + \"&downloadType=plain\"\n",
        "  url += \"&ids=\" + id\n",
        "  url += \"&upstreamAnchor=\" + upAnchor\n",
        "  url += \"&upstreamSign=\" + upSign + \"&upstreamOffset=\" + str(upOffs)\n",
        "  url += \"&downstreamAnchor=\" + downAnchor\n",
        "  url += \"&downstreamSign=\" + downSign + \"&downstreamOffset=\" + str(downOffs)\n",
        "  # retrieve from cache, if cached\n",
        "  if url in get_sequence_cache:\n",
        "    return get_sequence_cache[url]\n",
        "  # grab from url\n",
        "  response = urlopen(url)\n",
        "  string = response.read().decode(response.info().get_param(\"charset\") or \"utf-8-sig\")\n",
        "  result = \"\".join(string.splitlines()[1:])\n",
        "  # cache response and return\n",
        "  get_sequence_cache[url] = result\n",
        "  return result\n",
        "\n",
        "#@ _Fixed_ sequence retrieval from TriTrypDB (sequence requests spanning zero, eg. start codon minus 5 to plus 5, are broken and lose a base).\n",
        "\n",
        "def get_sequence(gene_id: str, feature: str, feature_end: str, indices: list):\n",
        "  \"\"\"\n",
        "  Fetch a sequence from TriTrypDB, behaviour as for sequence retrieval tool at https://tritrypdb.org/tritrypdb/app/fasta-tool/gene\n",
        "\n",
        "  :param id: Gene id\n",
        "  :param feature: Reference feature: `\"cds\"` for coding sequence, `\"transcript\"` for transcript.\n",
        "  :param end: Reference feature end: `\"start\"` for upstream end or `\"end\"` for downstream end.\n",
        "  :param start_index:\" Start of sequence relative to reference `feature` and `end`.\n",
        "  :param downOffs: End of sequence relative to reference `feature` and `end`.\n",
        "  :return: Sequence at the requested locus\n",
        "  \"\"\"\n",
        "  # dict for translating features into request anchors\n",
        "  anchor_translation = {\n",
        "      \"cds\": {\"start\": \"CodeStart\", \"end\": \"CodeEnd\"},\n",
        "      \"transcript\": {\"start\": \"Start\", \"end\": \"End\"}\n",
        "  }\n",
        "  # determine signs and offsets\n",
        "  signs = [None, None]\n",
        "  offs = [None, None]\n",
        "  for index, value in enumerate(indices):\n",
        "    if value >= 0:\n",
        "      signs[index] = \"plus\"\n",
        "      offs[index] = value\n",
        "    else:\n",
        "      signs[index] = \"minus\"\n",
        "      offs[index] = -value\n",
        "  offs[1] -= 1\n",
        "  if signs[1] == \"minus\":\n",
        "    offs[1] += 2\n",
        "  # get sequence, handling broken behaviour of requests spanning 0\n",
        "  seq = sequence_request(\n",
        "    gene_id,\n",
        "    anchor_translation[feature][feature_end], signs[0], offs[0],\n",
        "    anchor_translation[feature][feature_end], signs[1], offs[1]\n",
        "  )\n",
        "  # check response for problems\n",
        "  warnings = []\n",
        "  request_name = gene_id+\" \"+feature+\":\"+str(indices[0])+\"::\"+str(indices[1])\n",
        "  if \"N\" in seq.upper():\n",
        "    warnings.append(\"Ns in \" + request_name)\n",
        "    seq = seq.replace(\"N\", \"\")\n",
        "  if len(seq) != indices[1] - indices[0]:\n",
        "    warnings.append(\"Incorrect sequence length for \"+request_name)\n",
        "  # return result\n",
        "  if len(warnings) > 0:\n",
        "    return seq, warnings\n",
        "  else:\n",
        "    return seq, None\n",
        "\n",
        "#@ Reverse complement.\n",
        "\n",
        "def reverse_complement(seq) -> str:\n",
        "  \"\"\"\n",
        "  Reverse complements a DNA/RNA sequence, preserving case. U replaced with T.\n",
        "\n",
        "  :param seq: DNA or RNA sequence (without IUPAC degenerate bases)\n",
        "  :return: Reverse complemented sequence\n",
        "  \"\"\"\n",
        "  # replace U with T\n",
        "  replace = {\"U\": \"T\"}\n",
        "  for key in replace:\n",
        "    seq = seq.replace(key.lower(), replace[key].lower())\n",
        "  # define sequence complement\n",
        "  complement = {\"C\": \"G\", \"G\": \"C\", \"A\": \"T\", \"T\": \"A\"}\n",
        "  # update with lower case complement\n",
        "  lower_complement = {}\n",
        "  for key in complement:\n",
        "    lower_complement.update({key.lower(): complement[key].lower()})\n",
        "  complement.update(lower_complement)\n",
        "  # generate reverse complement sequence\n",
        "  rcseq = \"\"\n",
        "  for c in range(len(seq)):\n",
        "    rcseq = complement[seq[c]] + rcseq\n",
        "  return rcseq\n",
        "\n",
        "#@ Translation.\n",
        "\n",
        "# dna codon to single letter amino acid code dict\n",
        "codons = {\n",
        "  'TAA': '*', 'TAG': '*', 'TGA': '*',                                     # *\n",
        "  'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',                         # A\n",
        "  'TGT': 'C', 'TGC': 'C',                                                 # C\n",
        "  'GAT': 'D', 'GAC': 'D',                                                 # D\n",
        "  'GAA': 'E', 'GAG': 'E',                                                 # E\n",
        "  'TTT': 'F', 'TTC': 'F',                                                 # F\n",
        "  'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',                         # G\n",
        "  'CAT': 'H', 'CAC': 'H',                                                 # H\n",
        "  'ATT': 'I', 'ATC': 'I', 'ATA': 'I',                                     # I\n",
        "  'AAA': 'K', 'AAG': 'K',                                                 # K\n",
        "  'TTA': 'L', 'TTG': 'L', 'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L', # L\n",
        "  'ATG': 'M',                                                             # M\n",
        "  'AAT': 'N', 'AAC': 'N',                                                 # N\n",
        "  'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',                         # P\n",
        "  'CAA': 'Q', 'CAG': 'Q',                                                 # Q\n",
        "  'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R', # R\n",
        "  'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S', 'AGT': 'S', 'AGC': 'S', # S\n",
        "  'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',                         # T\n",
        "  'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',                         # V\n",
        "  'TGG': 'W',                                                             # W\n",
        "  'TAT': 'Y', 'TAC': 'Y'                                                  # Y\n",
        "}\n",
        "\n",
        "def translate(seq: str) -> str:\n",
        "  \"\"\"\n",
        "  Translates a DNA sequence to protein sequence using global variable `codons`.\n",
        "\n",
        "  :param seq: DNA sequence.\n",
        "  :return: Protein sequence, in whatever translation convention used by `codons`.\n",
        "  \"\"\"\n",
        "  pseq = \"\"\n",
        "  for codon_index in range(math.floor(len(seq) / 3)):\n",
        "    pseq += codons[seq[codon_index * 3:codon_index * 3 + 3]]\n",
        "  return pseq\n",
        "\n",
        "#@ PAM site identification.\n",
        "\n",
        "# list of dinucleotides which define pam\n",
        "pam_bases = [\"GG\", \"CC\"]\n",
        "\n",
        "def find_mutatable_pam(seq: str):\n",
        "  \"\"\"\n",
        "  Find PAM sites in a coding sequence which can be removed with a synonymous mutation.\n",
        "\n",
        "  :param seq: DNA sequence of CDS, must be in frame with a whole number of codons (ie. length is a multiple of 3).\n",
        "  :return: List of lists, recording pam position (start of NGG), direction (forward for NGG, reverse for CCN), mutation position and base for mutation\n",
        "  \"\"\"\n",
        "  # check length\n",
        "  warnings = []\n",
        "  if len(seq) % 3 != 0:\n",
        "    warnings.append(\"Sequence for PAM search not a multiple of 3 in length (unexpected behaviour may occur)\")\n",
        "  # result list of mutatable pams\n",
        "  result = []\n",
        "  # make query sequence uppercase\n",
        "  seq = seq.upper()\n",
        "  pseq = translate(seq)\n",
        "  # for each position in the sequence, if in a pam dinucleotide\n",
        "  for i in range(len(seq)):\n",
        "    for pam_offs in range(2):\n",
        "      if seq[i-pam_offs:i+2-pam_offs] in pam_bases:\n",
        "        # check all mutations at that position\n",
        "        for r in [\"A\", \"T\", \"C\", \"G\"]:\n",
        "          if r != seq[i]:\n",
        "            mutseq = seq[:i] + r + seq[i + 1:]\n",
        "            if translate(mutseq) == pseq:\n",
        "              # determine pam position and direction from pam_offs and if GG vs. CC\n",
        "              if seq[i] == \"G\":\n",
        "                pam_position = i - pam_offs - 1\n",
        "                pam_cutpos = pam_position - 4\n",
        "                pam_dir = \"forward\"\n",
        "              elif seq[i] == \"C\":\n",
        "                pam_position = i - pam_offs + 3\n",
        "                pam_cutpos = pam_position + 4\n",
        "                pam_dir = \"reverse\"\n",
        "              result.append([pam_position, pam_dir, pam_cutpos, i, seq[i], r])\n",
        "  if len(warnings) == 0:\n",
        "    warnings = None\n",
        "  return result, warnings\n",
        "\n",
        "def find_pam(seq: str):\n",
        "  \"\"\"\n",
        "  Find PAM sites in a DNA sequence and point mutations to remove them.\n",
        "\n",
        "  :param seq: DNA sequence of CDS, must be in frame with a whole number of codons (ie. length is a multiple of 3).\n",
        "  :return: List of lists, recording pam position (start of NGG), direction (forward for NGG, reverse for CCN), mutation position and base for mutation\n",
        "  \"\"\"\n",
        "  # for each position in the sequence, if in a pam dinucleotide\n",
        "  for i in range(len(seq)):\n",
        "    for pam_offs in range(2):\n",
        "      if seq[i-pam_offs:i+2-pam_offs] in pam_bases:\n",
        "        # check all mutations at that position\n",
        "        for r in [\"A\", \"T\", \"C\", \"G\"]:\n",
        "          if r != seq[i]:\n",
        "            # determine pam position and direction from pam_offs and if GG vs. CC\n",
        "            if seq[i] == \"G\":\n",
        "              pam_position = i - pam_offs - 1\n",
        "              pam_cutpos = pam_position - 4\n",
        "              pam_dir = \"forward\"\n",
        "            elif seq[i] == \"C\":\n",
        "              pam_position = i - pam_offs + 3\n",
        "              pam_cutpos = pam_position + 4\n",
        "              pam_dir = \"reverse\"\n",
        "            result.append([pam_position, pam_dir, pam_cutpos, i, seq[i], r])\n",
        "  return result\n",
        "\n",
        "\n",
        "# gene metadata check\n",
        "\n",
        "def check_gene_metadata(gene_id):\n",
        "  url = \"https://tritrypdb.org/tritrypdb/service/record-types/transcript/searches/GenesByText/reports/standard?text_search_organism=%5B%22Bodo%20saltans%20strain%20Lake%20Konstanz%22%2C%22Angomonas%20deanei%20strain%20Cavalho%20ATCC%20PRA-265%22%2C%22Blechomonas%20ayalai%20B08-376%22%2C%22Crithidia%20fasciculata%20strain%20Cf-Cl%22%2C%22Endotrypanum%20monterogeii%20strain%20LV88%22%2C%22Leishmania%20aethiopica%20L147%22%2C%22Leishmania%20amazonensis%20MHOM%2FBR%2F71973%2FM2269%22%2C%22Leishmania%20arabica%20strain%20LEM1108%22%2C%22Leishmania%20braziliensis%20MHOM%2FBR%2F75%2FM2903%22%2C%22Leishmania%20braziliensis%20MHOM%2FBR%2F75%2FM2904%22%2C%22Leishmania%20braziliensis%20MHOM%2FBR%2F75%2FM2904%202019%22%2C%22Leishmania%20donovani%20BPK282A1%22%2C%22Leishmania%20donovani%20CL-SL%22%2C%22Leishmania%20donovani%20HU3%22%2C%22Leishmania%20donovani%20strain%20LV9%22%2C%22Leishmania%20enriettii%20MCAV%2FBR%2F2001%2FCUR178%22%2C%22Leishmania%20enriettii%20strain%20LEM3045%22%2C%22Leishmania%20gerbilli%20strain%20LEM452%22%2C%22Leishmania%20infantum%20JPCM5%22%2C%22Leishmania%20major%20Friedlin%202021%22%2C%22Leishmania%20major%20strain%20Friedlin%22%2C%22Leishmania%20major%20strain%20LV39c5%22%2C%22Leishmania%20major%20strain%20SD%2075.1%22%2C%22Leishmania%20martiniquensis%20LEM2494%22%2C%22Leishmania%20martiniquensis%20MHOM%2FTH%2F2012%2FLSCM1%22%2C%22Leishmania%20mexicana%20MHOM%2FGT%2F2001%2FU1103%22%2C%22Leishmania%20orientalis%20MHOM%2FTH%2F2014%2FLSCM4%22%2C%22Leishmania%20panamensis%20MHOM%2FCOL%2F81%2FL13%22%2C%22Leishmania%20panamensis%20strain%20MHOM%2FPA%2F94%2FPSC-1%22%2C%22Leishmania%20sp.%20Ghana%20MHOM%2FGH%2F2012%2FGH5%22%2C%22Leishmania%20sp.%20Namibia%20MPRO%2FNA%2F1975%2F252%2FLV425%22%2C%22Leishmania%20tarentolae%20Parrot%20Tar%20II%202019%22%2C%22Leishmania%20tarentolae%20Parrot-TarII%22%2C%22Leishmania%20tropica%20L590%22%2C%22Leishmania%20turanica%20strain%20LEM423%22%2C%22Leptomonas%20pyrrhocoris%20H10%22%2C%22Leptomonas%20seymouri%20ATCC%2030220%22%2C%22Paratrypanosoma%20confusum%20CUL13%22%2C%22Porcisia%20hertigi%20MCOE%2FPA%2F1965%2FC119%22%2C%22Trypanosoma%20brucei%20EATRO1125%22%2C%22Trypanosoma%20brucei%20Lister%20strain%20427%22%2C%22Trypanosoma%20brucei%20Lister%20strain%20427%202018%22%2C%22Trypanosoma%20brucei%20brucei%20TREU927%22%2C%22Trypanosoma%20brucei%20gambiense%20DAL972%22%2C%22Trypanosoma%20congolense%20IL3000%22%2C%22Trypanosoma%20congolense%20IL3000%202019%22%2C%22Trypanosoma%20cruzi%20Berenice%22%2C%22Trypanosoma%20cruzi%20Brazil%20A4%22%2C%22Trypanosoma%20cruzi%20CL%20Brener%20Esmeraldo-like%22%2C%22Trypanosoma%20cruzi%20CL%20Brener%20Non-Esmeraldo-like%22%2C%22Trypanosoma%20cruzi%20Dm28c%202014%22%2C%22Trypanosoma%20cruzi%20Dm28c%202017%22%2C%22Trypanosoma%20cruzi%20Dm28c%202018%22%2C%22Trypanosoma%20cruzi%20Sylvio%20X10%2F1%22%2C%22Trypanosoma%20cruzi%20Sylvio%20X10%2F1-2012%22%2C%22Trypanosoma%20cruzi%20TCC%22%2C%22Trypanosoma%20cruzi%20Y%20C6%22%2C%22Trypanosoma%20cruzi%20marinkellei%20strain%20B7%22%2C%22Trypanosoma%20cruzi%20strain%20CL%22%2C%22Trypanosoma%20cruzi%20strain%20CL%20Brener%22%2C%22Trypanosoma%20cruzi%20strain%20G%22%2C%22Trypanosoma%20equiperdum%20OVI%22%2C%22Trypanosoma%20evansi%20strain%20STIB%20805%22%2C%22Trypanosoma%20grayi%20ANR4%22%2C%22Trypanosoma%20melophagium%20St.%20Kilda%22%2C%22Trypanosoma%20rangeli%20SC58%22%2C%22Trypanosoma%20theileri%20isolate%20Edinburgh%22%2C%22Trypanosoma%20vivax%20Y486%22%2C%22Eubodonida%22%2C%22Leishmania%20braziliensis%22%2C%22Leishmania%20donovani%22%2C%22Leishmania%20enriettii%22%2C%22Leishmania%20major%22%2C%22Leishmania%20martiniquensis%22%2C%22Leishmania%20panamensis%22%2C%22Leishmania%20tarentolae%22%2C%22Leptomonas%22%2C%22Trypanosoma%20brucei%22%2C%22Trypanosoma%20congolense%22%2C%22Trypanosoma%20cruzi%22%5D&text_expression=\"+gene_id+\"&document_type=gene&text_fields=%5B%22primary_key%22%5D&reportConfig=%7B%22attributes%22%3A%5B%22primary_key%22%2C%22cds_length%22%2C%22is_pseudo%22%2C%22signalp_peptide%22%2C%22gene_type%22%5D%2C%22tables%22%3A%5B%5D%2C%22attributeFormat%22%3A%22text%22%7D\"\n",
        "  response = urlopen(url)\n",
        "  string = response.read().decode(response.info().get_param(\"charset\") or \"utf-8-sig\")\n",
        "  import json\n",
        "  data = json.loads(string)\n",
        "  warnings = []\n",
        "  warnings_n = []\n",
        "  errors = [\"Gene ID not found\"]\n",
        "  for record in data[\"records\"]:\n",
        "    if record[\"displayName\"] == gene_id:\n",
        "      errors = []\n",
        "      if int(record[\"attributes\"][\"cds_length\"]) < minimum_homology_length * 2:\n",
        "        warnings.append(\"Very short CDS may lead to misplaced sgRNAs\")\n",
        "      if record[\"attributes\"][\"signalp_peptide\"] is not None:\n",
        "        warnings_n.append(\"SignalP-predicted signal peptide \"+str(record[\"attributes\"][\"signalp_peptide\"])+\" may prevent N terminal tagging\")\n",
        "      if record[\"attributes\"][\"is_pseudo\"] != \"No\":\n",
        "        warnings.append(\"Listed as a pseudogene\")\n",
        "  if len(warnings) == 0:\n",
        "    warnings = None\n",
        "  if len(warnings_n) == 0:\n",
        "    warnings_n = None\n",
        "  if len(errors) == 0:\n",
        "    errors = None\n",
        "  return errors, warnings, warnings_n"
      ],
      "metadata": {
        "id": "5anOTzfcjyyg",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title User settings\n",
        "\n",
        "#@markdown List of gene IDs for which to design primers\n",
        "# as in TriTrypDB\n",
        "gene_ids = \"Tb927.5.3210 Tb427_050038600 \" #@param {type:\"string\"}\n",
        "gene_ids = gene_ids.split()\n",
        "\n",
        "#@markdown Tagging terminus\n",
        "# \"n\", \"c\" or \"both\"\n",
        "terminus = \"both\" #@param string [\"n\", \"c\", \"both\"]\n",
        "\n",
        "#@markdown Plasmid system, only 2A gives scarless integration\n",
        "# defines primer binding site sequences\n",
        "plasmid_system = \"ppot-compatible\" #@param string [\"ppot-compatible\", \"2a\"]\n",
        "\n",
        "#@markdown Tag and drug marker (when using 2A plasmid system)\n",
        "tag = \"mng\"  #@param string [\"msc\", \"mng\"]\n",
        "drug = \"bsr\" #@param string [\"pac\", \"bsr\"]\n",
        "\n",
        "# guide rna length (excluding pam ngg), bases\n",
        "guide_length = 20\n",
        "\n",
        "# minimum homology arm length, bases\n",
        "minimum_homology_length = 30\n",
        "\n",
        "# maximum primer length, bases\n",
        "maximum_primer_length = 100\n",
        "\n",
        "# primer binding sites\n",
        "if plasmid_system == \"ppot-compatible\":\n",
        "  binding_n_forward = \"gtataatgcagacctgctgc\".lower()\n",
        "  binding_n_reverse = \"ctggatcaggatcgggtagt\".lower()\n",
        "  binding_c_forward = \"ggttctggtagtggttccgg\".lower()\n",
        "  binding_c_reverse = \"gcacaggtctctcaaattgg\".lower()\n",
        "  print(\"Warning: This primer design does not give scarless integration\")\n",
        "elif plasmid_system == \"2a\":\n",
        "  if tag == \"mng\":\n",
        "    binding_n_reverse = \"ATGGACGAGCTGTACAAG\".lower()\n",
        "    binding_c_forward = \"GTGTCCAAGGGCGAGGAG\".lower()\n",
        "  if tag == \"msc\":\n",
        "    binding_n_reverse = \"atggacgagctgtacaag\".lower()\n",
        "    binding_c_forward = \"GTGAGCAAGGGCGAGGCA\".lower()\n",
        "  if drug == \"bsr\":\n",
        "    binding_n_forward = \"ATGGCCAAGCCGCTCAGC\".lower()\n",
        "    binding_c_reverse = \"TACGTCTGGGAGGGCTAG\".lower()\n",
        "  if drug == \"pac\":\n",
        "    binding_n_forward = \"ATGACGGAGTACAAACCC\".lower()\n",
        "    binding_c_reverse = \"CGCAAGCCGGGCGCGTGA\".lower()\n",
        "\n",
        "sgrna_t7 = \"gaaattaatacgactcactatagg\".lower()\n",
        "sgrna_hairpin = \"gttttagagctagaaatagc\".lower()\n",
        "\n",
        "verbose = False"
      ],
      "metadata": {
        "id": "tpa0pcE2jvoo",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Find tagging primers\n",
        "\n",
        "#@markdown Exhaustively scans the start (for N terminal tagging) or end (for C terminal tagging) of the open reading frame for PAM sites which can be mutated to remove the PAM without changing the coding sequence.\n",
        "\n",
        "#@markdown Up to six types of primers are designed:\n",
        "\n",
        "def print_cds_pam_result(seq, pam_results):\n",
        "\tprint(\"PAM search area \", seq)\n",
        "\tprint(\"Protein sequence\", translate(seq))\n",
        "\tprint(\"Results:\", \"       \", \"(\"+\", \".join([\"PAM posn.\", \"PAM dir.\", \"Cut posn.\", \"Mut. posn.\", \"Orig. base\", \"Mut. base\"])+\")\")\n",
        "\tfor result in pam_results:\n",
        "\t\tprint(\"\", \", \".join([str(x) for x in result]))\n",
        "\n",
        "primer_results = []\n",
        "\n",
        "for gene_id in gene_ids:\n",
        "  primer_result = {\"id\": gene_id, \"uf\": None, \"ur\": None, \"usg\": None, \"uerr\": None, \"df\": None, \"dr\": None, \"dsg\": None, \"derr\": None}\n",
        "  gene_errors, gene_warnings, gene_warnings_n = check_gene_metadata(gene_id)\n",
        "  if gene_errors is not None:\n",
        "    primer_result[\"uerr\"] = gene_errors\n",
        "    primer_result[\"derr\"] = gene_errors\n",
        "  else:\n",
        "    if terminus == \"n\" or terminus==\"both\":\n",
        "      #@markdown For N terminal tagging, gives an upstream forward (UF), upstream reverse (UR) and upstream sgRNA (USG) primer\n",
        "      # find pams in cds, start of protein coding sequence\n",
        "      if gene_warnings is not None:\n",
        "        warnings = gene_warnings.copy() + gene_warnings_n.copy()\n",
        "      else:\n",
        "        warnings = []\n",
        "      pam_search_range = 3 * math.floor((maximum_primer_length - len(binding_n_reverse) - minimum_homology_length) / 3)\n",
        "      seq, warning = get_sequence(gene_id, \"cds\", \"start\", [3, 3 + pam_search_range])\n",
        "      if warning: warnings += warning\n",
        "      pam_results, warning = find_mutatable_pam(seq)\n",
        "      if warning: warnings += warning\n",
        "      if verbose: print_cds_pam_result(seq, pam_results)\n",
        "      if len(pam_results) == 0:\n",
        "        # if no sites found, fail!\n",
        "        primer_result[\"uerr\"] = [\"No mutatable PAM sites found near start of CDS\"]\n",
        "      else:\n",
        "        # first in list, closest to n terminus\n",
        "        pam_result = pam_results[0]\n",
        "        # upstream forward primer\n",
        "        upstream_forward_sequence, warning = get_sequence(gene_id, \"cds\", \"start\", [-minimum_homology_length, 0]) # Removed offset of -1\n",
        "        if warning: warnings += warning\n",
        "        upstream_forward = upstream_forward_sequence + binding_n_forward\n",
        "        if verbose: print(\"Upstream forward:  \", upstream_forward)\n",
        "        primer_result[\"uf\"] = upstream_forward\n",
        "        # upstream reverse primer\n",
        "        reverse_homology, warning = get_sequence(gene_id, \"cds\", \"start\", [3, 3 + minimum_homology_length + pam_result[2]])\n",
        "        if warning: warnings += warning\n",
        "        reverse_homology = reverse_homology[:pam_result[3]] + pam_result[5].lower() + reverse_homology[pam_result[3] + 1:]\n",
        "        upstream_reverse = reverse_complement(binding_n_reverse + reverse_homology)\n",
        "        if verbose: print(\"Upstream reverse:  \", upstream_reverse)\n",
        "        primer_result[\"ur\"] = upstream_reverse\n",
        "        # upstream sgrna primer\n",
        "        if pam_result[1] == \"forward\":\n",
        "          upstream_sgrna_sequence, warning = get_sequence(gene_id, \"cds\", \"start\", [3 + pam_result[0] - guide_length, 3 + pam_result[0]])\n",
        "          if warning: warnings += warning\n",
        "          upstream_sgrna = sgrna_t7 + upstream_sgrna_sequence + sgrna_hairpin\n",
        "        elif pam_result[1] == \"reverse\":\n",
        "          upstream_sgrna_sequence, warning = get_sequence(gene_id, \"cds\", \"start\", [3 + pam_result[0], 3 + pam_result[0] + guide_length])\n",
        "          if warning: warnings += warning\n",
        "          upstream_sgrna = sgrna_t7 + reverse_complement(upstream_sgrna_sequence) + sgrna_hairpin\n",
        "        if verbose: print(\"Upstream sgRNA:    \", upstream_sgrna)\n",
        "        primer_result[\"usg\"] = upstream_sgrna\n",
        "        if len(warnings) > 0:\n",
        "          primer_result[\"uerr\"] = warnings\n",
        "    if terminus == \"c\" or terminus==\"both\":\n",
        "      #@markdown For C terminal tagging, gives a downstream forward (DF), downstream reverse (DR) and downstream sgRNA (DSG) primer\n",
        "      # find pam in cds, end of protein coding sequence, minus stop codon\n",
        "      if gene_warnings is not None:\n",
        "        warnings = gene_warnings.copy()\n",
        "      else:\n",
        "        warnings = []\n",
        "      pam_search_range = 3 * math.floor((maximum_primer_length - len(binding_c_forward) - minimum_homology_length) / 3)\n",
        "      seq, warning = get_sequence(gene_id, \"cds\", \"end\", [-pam_search_range - 3, -3])\n",
        "      if warning: warnings += warning\n",
        "      pam_results, warning = find_mutatable_pam(seq)\n",
        "      if warning: warnings += warning\n",
        "      if verbose: print_cds_pam_result(seq, pam_results)\n",
        "      if len(pam_results) == 0:\n",
        "        #if no sites found, fail!\n",
        "        primer_result[\"derr\"] = [\"No mutatable PAM sites found near end of CDS\"]\n",
        "      else:\n",
        "        # last in list, closest to c terminus\n",
        "        pam_result = pam_results[-1]\n",
        "        # pam sites were found in forward direction, so adjust with pam_search_range\n",
        "        # downstream forward primer\n",
        "        forward_homology, warning = get_sequence(gene_id, \"cds\", \"end\", [-minimum_homology_length - (pam_search_range - pam_result[2]) - 3 + 1, - 3 + 1]) # Added offset of +1 to correct for gene end indexing\n",
        "        if warning: warnings += warning\n",
        "        forward_homology = forward_homology[:-pam_search_range + pam_result[3]] + pam_result[5].lower() + forward_homology[-pam_search_range + pam_result[3] + 1:] # TODO? Is an offset of +1 also required here? Is PAM site correct in position?\n",
        "        downstream_forward = forward_homology + binding_c_forward\n",
        "        if verbose: print(\"Downstream forward:\", downstream_forward)\n",
        "        primer_result[\"df\"] = downstream_forward\n",
        "        # downstream reverse primer\n",
        "        downstream_reverse_sequence, warning = get_sequence(gene_id, \"cds\", \"end\", [0 + 1, minimum_homology_length + 1])\n",
        "        if warning: warnings += warning\n",
        "        downstream_reverse = reverse_complement(binding_c_reverse + downstream_reverse_sequence)\n",
        "        if verbose: print(\"Downstream reverse:\", downstream_reverse)\n",
        "        primer_result[\"dr\"] = downstream_reverse\n",
        "        # downstream sgrna primer\n",
        "        if pam_result[1] == \"forward\":\n",
        "          downstream_sgrna_sequence, warning = get_sequence(gene_id, \"cds\", \"end\", [-pam_search_range + pam_result[0] - guide_length - 3, -pam_search_range + pam_result[0] - 3])\n",
        "          if warning: warnings += warning\n",
        "          downstream_sgrna = sgrna_t7 + downstream_sgrna_sequence + sgrna_hairpin\n",
        "        elif pam_result[1] == \"reverse\":\n",
        "          downstream_sgrna_sequence, warning = get_sequence(gene_id, \"cds\", \"end\", [-pam_search_range + pam_result[0] - 3 + 1, -pam_search_range + pam_result[0] + guide_length + 1]) # Added offset of +1 to correct for gene end indexing\n",
        "          if warning: warnings += warning\n",
        "          downstream_sgrna = sgrna_t7 + reverse_complement(downstream_sgrna_sequence) + sgrna_hairpin\n",
        "        if verbose: print(\"Downstream sgRNA:  \", downstream_sgrna)\n",
        "        primer_result[\"dsg\"] = downstream_sgrna\n",
        "        if len(warnings) > 0:\n",
        "          primer_result[\"derr\"] = warnings\n",
        "  primer_results.append(primer_result)\n",
        "\n",
        "file = open(\"primers.tsv\", \"w\")\n",
        "file.write(\"\\t\".join([\"Gene ID\", \"UF\", \"UR\", \"USG\", \"UERR\", \"DF\", \"DR\", \"DSG\", \"DERR\"])+\"\\r\\n\")\n",
        "for result in primer_results:\n",
        "  for err in [\"uerr\", \"derr\"]:\n",
        "    if result[err] is None:\n",
        "      result[err] = \"None\"\n",
        "    else:\n",
        "      result[err] = \", \".join(result[err])\n",
        "  for key in result:\n",
        "    print(\"\\t\".join([str(x) for x in [key, result[key]]]))\n",
        "  print(\"\")\n",
        "  file.write(\"\\t\".join([str(x) for x in [result[\"id\"], result[\"uf\"], result[\"ur\"], result[\"usg\"], result[\"uerr\"], result[\"df\"], result[\"dr\"], result[\"dsg\"], result[\"derr\"]]])+\"\\r\\n\")\n",
        "file.close()\n",
        "\n",
        "from google.colab import files\n",
        "files.download(\"primers.tsv\")\n",
        "\n",
        "#@markdown Sequences which are shared among all primers of that type (ie. standard plasmid annealing sequences, sgRNA T7 promoter sequence) are shown in lower case.\n",
        "#@markdown Gene/terminus-specific sequences (homology arms and actual guide RNA) are shown in upper case.\n",
        "#@markdown The single nucleotide change to mutate the PAM site is shown in lowercase, in the upstream reverse primer for N terminal tagging, and the downstream reverse primer for C terminal tagging.\n",
        "\n",
        "#@markdown Make sure to check the upstream error (UERR) and downstream error (DERR) columns for any errors or warnings about the primer design."
      ],
      "metadata": {
        "id": "JSj3u9aVl5Q5",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}